src/GithubWebhookJira/Webhook.php

```diff
@@ -0,0 +1,276 @@
+      'jiraPassword' => getenv('JIRA_PASSWORD'),
+      'jiraLogLevel' => 'EMERGENCY',
+    ]);
+    $this->_issue = new IssueService($config);
+
+    $this->_raw_data = $request->getContent();
+  }
+
+  /**
+   * Check if the webhook is a valid request
+   *
+   * @return boolean
+   */
+  public function isValid() : bool {
+    list($algo, $sig) = explode(
+      '=',
+      $this->_request->headers->get('X-Hub-Signature')
+    );
+    return hash_hmac($algo, $this->_raw_data, $this->_secret) === $sig;
+  }
+
+  /**
+   * Get our data
+   *
+   * @return \stdClass
+   */
+  private function _getData() : \stdClass {
+    return json_decode($this->_raw_data);
+  }
+
+  /**
+   * Process our webhook data
+   *
+   * @return void
+   */
+  public function process() {
+    switch ($this->_getData()->action) {
+      case self::ACTION_REOPENED:
+        /* falls through */
+      case self::ACTION_EDITED:
+        /* falls through */
+      case self::ACTION_OPENED:
+        $this->_processPullRequestOpen();
+        break;
+      case self::ACTION_CLOSED:
+        $this->_processPullRequestClose();
+        break;
+      default:
+        // do nothing
+    }
+  }
+
+  /**
+   * Handle opened PR action
+   */
+  private function _processPullRequestOpen() {
+    $items = $this->_getJiraItems();
+    $check = array_filter($items, function($item) {
+      return empty($item['url']);
+    });
+
+    if ( !empty($check)) {
+      $this->_updateUrls($check);
+    }
+
+    foreach ($items as $item) {
+      $transition = new Transition();
+      $transition->setTransitionId(self::TRANSITION_PEER_REVIEW);
+      $this->_issue->transition($item['key'], $transition);
+      $comment = new Comment();
+      $comment->setBody(
+        "PR Opened: [{$this->_getData()->repository->full_name}#" .
+        "{$this->_getData()->pull_request->number}|" .
+        "{$this->_getData()->pull_request->html_url}]"
+      );
+      $this->_issue->addComment($item['key'], $comment);
+    }
+  }
+
+  /**
+   * Handle closed PR action
+   */
+  private function _processPullRequestClose() {
+    if ($this->_getData()->pull_request->merged === true) {
+      foreach ($this->_getJiraItems() as $item) {
+        $transition = new Transition();
+        $transition->fields['resolution'] = ['name' => self::RESOLUTION_DONE];
+        $transition->setTransitionId(self::TRANSITION_DONE);
+        $this->_issue->transition($item['key'], $transition);
+      }
+    } else {
+      foreach ($this->_getJiraItems() as $item) {
+        $transition = new Transition();
+        $transition->setTransitionId(self::TRANSITION_REVIEW_REJECTED);
+        $this->_issue->transition($item['key'], $transition);
+      }
+    }
+  }
+
+  /**
+   * Update the Jira IDs with URLs in the pull request
+   */
+  private function _updateUrls() {
+    $regex =
+      '((?:(close|closes|closed|fix|fixes|fixed|resolve|resolves|resolved))' .
+      '\s(?!' .
+      $this->_jira_url .
```

Comment:
if it's a url already, nothing really needs to happen

Reply:
The exclamation before it means NOT. This doesn't process anything that matches a url

Reply:
ah. i missed that. but is all that needed? `fixes ABC-123` will never match a url if it's followed by whitespace or end of line

Reply:
Good point, will fix.